magnum-zio/src/main/scala/com/augustnagro/magnum/magzio/Transactor.scala

```diff
@@ -0,0 +1,173 @@
+package com.augustnagro.magnum.magzio
+
+import com.augustnagro.magnum.{DbCon, DbTx, SqlException, SqlLogger}
+import zio.{IO, Semaphore, Task, Trace, UIO, URIO, ZIO}
+
+import java.sql.Connection
+import javax.sql.DataSource
+import scala.reflect.ClassTag
+import scala.util.control.{ControlThrowable, NonFatal}
+
+class Transactor private (
+    dataSource: DataSource,
+    sqlLogger: SqlLogger,
+    connectionConfig: Connection => Unit,
+    semaphore: Option[Semaphore]
+):
+
+  def withSqlLogger(sqlLogger: SqlLogger): Transactor =
+    new Transactor(
+      dataSource,
+      sqlLogger,
+      connectionConfig,
+      semaphore
+    )
+
+  def withConnectionConfig(connectionConfig: Connection => Unit): Transactor =
+    new Transactor(
+      dataSource,
+      sqlLogger,
+      connectionConfig,
+      semaphore
+    )
+
+  def connect[A](f: DbCon ?=> A)(using Trace): Task[A] =
+    val zio = ZIO.blocking(
+      ZIO.acquireReleaseWith(acquireConnection)(releaseConnection)(cn =>
+        ZIO.attempt {
+          connectionConfig(cn)
+          f(using DbCon(cn, sqlLogger))
+        }
+      )
+    )
+    semaphore.fold(zio)(_.withPermit(zio))
+
+  def transact[A](f: DbTx ?=> A)(using Trace): Task[A] =
+    val zio = ZIO.blocking(
+      ZIO.acquireReleaseWith(acquireConnection)(releaseConnection)(cn =>
+        ZIO.attempt {
+          connectionConfig(cn)
+          cn.setAutoCommit(false)
+          try
+            val res = f(using DbTx(cn, sqlLogger))
+            cn.commit()
+            res
+          catch
+            case NonFatal(t) =>
+              cn.rollback()
+              throw t
+        }.uninterruptible
+      )
+    )
+    semaphore.fold(zio)(_.withPermit(zio))
+
+  private def acquireConnection(using Trace): Task[Connection] =
+    ZIO
+      .attempt(dataSource.getConnection())
+      .mapError(t => SqlException("Unable to acquire DB Connection", t))
+
+  private def releaseConnection[R](con: Connection)(using Trace): URIO[R, Any] =
```

Comment:
Can be simplified. There's no `R` here
```suggestion
  private def releaseConnection(con: Connection)(using Trace): UIO[Unit]
```

Reply:
fixed in your MR.